python/ray/tune/impl/utils.py

```diff
@@ -0,0 +1,29 @@
+from ray.data import Dataset
+from ray.tune.sample import Categorical
+
+
+def execute_dataset(config_dict: dict):
+    """Going through config dict (params space) and fully execute any Dataset
+    if necessary.
+    """
+    for k, v in config_dict.items():
+        if isinstance(v, dict):
+            execute_dataset(v)
+        elif isinstance(v, Dataset):
+            config_dict[k] = v.fully_executed()
+        # TODO(xwjiang): Consider CV config for beta.
+        # elif isinstance(v, int):
+        #     # CV settings
+        #     pass
+        elif isinstance(v, list) or isinstance(v, Categorical):
+            _list = v if isinstance(v, list) else v.categories
+            if len(_list) == 0:
+                return
+            assert all(isinstance(_item, type(_list[0])) for _item in _list)
```

Comment:
This is a minor nit, but we don't require this constraint in search space definition

Reply:
Let's ignore this for now though

Reply:
ah good to know! We should add that case to our unit test/examples at some time.